### PR TITLE
[316 / several] Add group and membership related emails

### LINKED ISSUE #187: Group destruction should notify all members
Woo more e-mails

### PR DESCRIPTION
**Pull request by esoterik**
Resolves #187, #239, #316
 - Add notification to group members on disband
 - Add member notification to GroupFinalizer
 - Add email when a group locks
 - Add email to leader when accepted memberships change

### CODE REVIEW

**Comment by esoterik** [comment]
blocked on #397

File: app/mailers/student_mailer.rb [hunk 1]
```diff
@@ -21,7 +21,44 @@ def draw_invitation(user, college)
   # @param college [College] the college to pull settings from
   def selection_invite(user, college)
     @user = user
-    @college = college
+    @college = college || College.new
     mail(to: @user.email, subject: 'Time to select a suite!')
   end
+
+  def disband_notification(user, college)
```
Comment (esoterik) [inline app/mailers/student_mailer.rb]: these all need documentation

File: app/mailers/student_mailer.rb [hunk 2]
```diff
@@ -11,7 +11,7 @@ class StudentMailer < ApplicationMailer
   def draw_invitation(user, college)
     @user = user
     @intent_deadline = user.draw.intent_deadline
-    @college = college
+    @college = college || College.new
```
Comment (orenyk) [inline app/mailers/student_mailer.rb]: Can this be extracted somehow? It's _super_ repetitive as-is.

File: app/mailers/student_mailer.rb [hunk 3]
```diff
@@ -21,7 +21,44 @@ def draw_invitation(user, college)
...
+                         groups_url(user.group)
+                       end
+    mail(to: @user.email, subject: 'Your housing group is now finalizing')
+  end
+
+  def joined_group(joined:, group:, college:)
+    @user = group.leader
+    @joined = joined
+    @college = college || College.new
+    mail(to: @user.email, subject: "#{joined.full_name} has joined your group!")
```
Comment (orenyk) [inline app/mailers/student_mailer.rb]: I don't think we need the exclamation point

File: app/mailers/student_mailer.rb [hunk 4]
```diff
@@ -21,7 +21,44 @@ def draw_invitation(user, college)
...
+
+  def finalizing_notification(user, college)
+    @user = user
+    @college = college || College.new
+    @finalizing_path = if user.group.draw
+                         draw_groups_url(user.group)
+                       else
+                         groups_url(user.group)
+                       end
+    mail(to: @user.email, subject: 'Your housing group is now finalizing')
```
Comment (orenyk) [inline app/mailers/student_mailer.rb]: I think we can change this to "Please confirm your housing group" --> I think we're getting rid of the 'finalizing' language anyway in #352

File: app/mailers/student_mailer.rb [hunk 5]
```diff
@@ -21,7 +21,44 @@ def draw_invitation(user, college)
...
+  end
+
+  def joined_group(joined:, group:, college:)
```
Comment (orenyk) [inline app/mailers/student_mailer.rb]: Lol you switched to named params halfway through - let's just pick one and stick to it.

File: app/mailers/student_mailer.rb (hunk 2 above)
Comment (orenyk) [inline app/mailers/student_mailer.rb]: Also, it occurs to me that we don't need to pass the college to the mailer methods at this stage - we can just set it to `College.first || College.new` globally. On the flip side, we'll eventually want to pass in a college (unless we just want to inherit it from the user?) - I don't feel strongly either way so it's your call.

File: app/views/student_mailer/disband_notification.html.erb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html>
```
Comment (orenyk) [inline app/views/student_mailer/disband_notification.html.erb]: I know it's annoying, and we'll definitely want to refactor this (opening up an issue now), but can we include both HTML and text versions of all e-mails?

File: app/views/student_mailer/disband_notification.html.erb [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+<html>
+  <head>
+    <meta content='text/html; charset=UTF-8' http-equiv='Content-Type' />
+  </head>
+  <body>
+    <p>Hi <%= @user.name %>,</p>
+    <p>Your housing group has been disbanded, either by your leader or an admin.</p>
+    <p>Please e-mail <%= @college.admin_email %> if you have any questions.</p>
```
Comment (orenyk) [inline app/views/student_mailer/disband_notification.html.erb]: Does this need to be a separate paragraph? I'd just combine both sentences into one.

File: app/views/student_mailer/finalizing_notification.html.erb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta content='text/html; charset=UTF-8' http-equiv='Content-Type' />
+  </head>
+  <body>
+    <p>Hi <%= @user.name %>,</p>
+    <p>Your group is now in the process of finalizing! Please go to <%= "#{@college.site_url} #{@finalizing_path}" %> to confirm your membership.</p>
```
Comment (orenyk) [inline app/views/student_mailer/finalizing_notification.html.erb]: Let's reword this to include language about "confirming the housing group" (see my earlier comment and #352)

File: app/views/student_mailer/disband_notification.html.erb (hunk 7 above)
Comment (orenyk) [inline app/views/student_mailer/disband_notification.html.erb]: This is true across all of these e-mails. The first sentence / part is short enough in all of them that they should just be consolidated.

File: app/views/student_mailer/joined_group.html.erb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta content='text/html; charset=UTF-8' http-equiv='Content-Type' />
+  </head>
+  <body>
+    <p>Hi <%= @user.name %>,</p>
+    <p><%= @joined.full_name %> has joined your housing group! %>
```
Comment (orenyk) [inline app/views/student_mailer/joined_group.html.erb]: Missing closing p tag.

File: spec/models/group_spec.rb [hunk 10]
```diff
@@ -228,4 +228,26 @@ def create_potential_member(status:, group:)
       expect(group).not_to be_unlockable
     end
   end
+
+  context 'on disbanding' do
+    let(:msg) { instance_spy(ActionMailer::MessageDelivery, deliver_later: 1) }
```
Comment (orenyk) [inline spec/models/group_spec.rb]: Nice, I like this stub better than how we were doing it before 😄.

File: spec/models/user_spec.rb [hunk 11]
```diff
@@ -42,10 +42,10 @@
   # rubocop:disable RSpec/ExampleLength
   it 'destroys a dependent membership on destruction' do
     user = FactoryGirl.create(:student, intent: 'on_campus')
-    FactoryGirl.create(:drawless_group, leader: user)
-    membership_id = user.membership.id
+    g = FactoryGirl.create(:drawless_group, size: 2)
```
Comment (orenyk) [inline spec/models/user_spec.rb]: Isn't this in #410?

File: spec/services/group_finalizer_spec.rb [hunk 12]
```diff
@@ -12,9 +12,18 @@
           .and_return(instance_spy('Draw', open_suite_sizes: [g.size]))
         membership = instance_spy('Membership', locked: false)
         leader = instance_spy('User', membership: membership)
+        m = Array.new(g.size - 1) { |_| instance_spy('user') }
```
Comment (orenyk) [inline spec/services/group_finalizer_spec.rb]: Whew this let block is getting ugly - maybe we extract to a `mock_group` method?

File: app/mailers/student_mailer.rb [hunk 13]
```diff
@@ -9,9 +9,9 @@ class StudentMailer < ApplicationMailer
   # @param user [User] the user to send the invitation to
   # @param college [College] the college to pull settings from
   def draw_invitation(user, college)
```
Comment (orenyk) [inline app/mailers/student_mailer.rb]: These two methods still aren't using named params

File: app/mailers/student_mailer.rb [hunk 14]
```diff
@@ -20,45 +20,75 @@ def draw_invitation(user, college)
...
+    determine_college(college)
     @user = user
-    @college = college || College.new
     mail(to: @user.email, subject: 'Your housing group is now locked')
   end
+
+  private
+
+  def determine_college(college)
+    @college = if college
```
Comment (orenyk) [inline app/mailers/student_mailer.rb]: can't this just be `@college = college || College.first || College.new`?

File: app/views/student_mailer/disband_notification.html.erb [hunk 15]
```diff
@@ -6,7 +6,7 @@
   <body>
     <p>Hi <%= @user.name %>,</p>
     <p>Your housing group has been disbanded, either by your leader or an admin.</p>
-    <p>Please e-mail <%= @college.admin_email %> if you have any questions.</p>
+    Please e-mail <%= @college.admin_email %> if you have any questions.
```
Comment (orenyk) [inline app/views/student_mailer/disband_notification.html.erb]: This needs to be inside the `<p>` tags on the previous line.

File: app/views/student_mailer/finalizing_notification.html.erb [hunk 16]
```diff
@@ -5,7 +5,7 @@
   </head>
   <body>
     <p>Hi <%= @user.name %>,</p>
-    <p>Your group is now in the process of finalizing! Please go to <%= "#{@college.site_url} #{@finalizing_path}" %> to confirm your membership.</p>
+    <p>Your group is now in the process of being confirmed. Please go to <%= "#{@college.site_url} #{@finalizing_path}" %> to confirm your membership.</p>
```
Comment (orenyk) [inline app/views/student_mailer/finalizing_notification.html.erb]: Won't the space between the url and path cause problems here? Also, I think we can use `<a>` tags for links in the HTML e-mail (e.g. `<a href="<%= @college.site_url + finalizing_path %>Vesta</a>`).

File: app/views/student_mailer/finalizing_notification.text.erb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+Hi <%= @user.name %>,
+Your group is now in the process of being confirmed. Please go to <%= "#{@college.site_url} #{@finalizing_path}" %> to confirm your membership.
```
Comment (orenyk) [inline app/views/student_mailer/finalizing_notification.text.erb]: Same comment re: space between site url and path.

File: app/views/student_mailer/group_locked.html.erb [hunk 18]
```diff
@@ -5,8 +5,8 @@
   </head>
   <body>
     <p>Hi <%= @user.name %>,</p>
-    <p>Your housing group is now locked.</p>
-    <p>Please e-mail <%= @college.admin_email %> if you need to make any changes.</p>
+    <p>Your housing group is now locked.
+    Please e-mail <%= @college.admin_email %> if you need to make any changes.</p>
```
Comment (orenyk) [inline app/views/student_mailer/group_locked.html.erb]: Can we make this one line?
